WordPress/gutenberg
# Add new `@wordpress/preferences` package

## Description
First part of https://github.com/WordPress/gutenberg/issues/31965

This PR contains part of the code from #38435, I've separated it out for easier review and will keep the two in sync.

It adds a new `@wordpress/preferences` package with a data store and a few components.

Here's some links to the documentation for everything this PR introduces:
- [Main package README](https://github.com/WordPress/gutenberg/blob/64ed5098a09796a60a6bd47a618e55a4b160aef9/packages/preferences/README.md) (which includes the action/selector documentation and some brief example code)
- [`PreferenceToggleMenuItem` component docs](https://github.com/WordPress/gutenberg/blob/64ed5098a09796a60a6bd47a618e55a4b160aef9/packages/preferences/src/components/preference-toggle-menu-item/README.md)

This is a good time to review the API and do any renaming. At the moment, everything works similarly to the way it does in `trunk`, but there is scope now to change things.

Currently the only preferences managed in the new package are what we call 'feature' preferences, which are just simple boolean flags. Surprisingly this covers almost all the preferences in our editors, with the exception of a couple of things:
- The code editor/visual editor toggle (this would be more of an enum type setting, and can be handled in a follow-up)
- The block manager (which I think should probably be kept entirely separate to preferences because it relies on block registration and block types)

## Testing Instructions
For testing use the `add/new-preferences-package` branch from #38435. Everything should behave the same as `trunk` in the post editor, widgets editor and customize widgets editor which I've refactored in that PR to use the new preferences store.

## Types of changes


New package (non-breaking change which adds functionality)


## Checklist:
